EIPS/eip-4804.md

```diff
@@ -0,0 +1,140 @@
+
+### Resolve Mode
+
+Once the "To" address and chainid are determined, the protocol will check the resolver mode of contract by calling "resolveMode" method. The protocol currently supports two resolve modes:
+
+#### Manual Mode
+
+The manual mode will not do any interpretation of **path**, and put **path** as calldata of the message directly.
+
+#### Auto Mode
+
+The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Ethereum contract ABI [Contract ABI Specification — Solidity 0.8.3 documentation](https://docs.soliditylang.org/en/v0.8.3/abi-spec.html), where
+
+- **method** is a string of function method be called
+- **argument_i** is the ith argument of the method. If **type** is specified, the value will be translated to the corresponding type. The protocol currently supports the basic types such as uint256, bytes32, address, and bytes. If **type** is not specified, then the type will be automatically detected using the following rule in a sequential way:
+
+1. **type** = "uint256", if **value** is numeric; or
+2. **type**="bytes32", if **value** is in the form of 0x+32-byte-data hex; or
+3. **type**="address", if **value** is in the form of 0x+20-byte-data hex; or
+4. **type**="name", if **value** is in the form of **name**.**nsProvider**. In this case, the actual value of the argument will be **address**, which is obtained from **nsProvider**, e.g., eth, w3q, etc.
+5. else **type**="bytes"
+
+Note that if **method** does not exist, i.e., **path** is empty or "/", then the contract will be called with empty calldata.
+
+## Examples
+
+#### Example 1
+
+```
+web3://ensdomains.eth:4
+```
+
+The protocol will find the address of **ensdomains.eth** from ENS in chainid 4 (Rinkeby), and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 4.
+
+#### Example 2
+
+```
+web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
+```
+
+The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.
+
+#### Example 3
+
+```
+web3://cyberbrokers-meta.eth:1/renderBroker/9999
+```
+
+The protocol will find the address of **cyberbrokers-meta.eth** from ENS, and then call the address with "From" = "0x..." and "Calldata" = "0x" + `keccak("view(uint256)")[0:4] + abi.encode(uint256(9999))` with chainid = 1.
+
+#### Exmaple 4
+
+```
+web3://home.w3q:3334/
+```
+
+The protocol will find the address of **home.w3q** from W3NS, and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 3334.
```

Comment:
I think this EIP is going to need a list of `nsProvider` suffixes, and how to resolve them.

Reply:
A section about how to solve is added.  Please take a look.